// jovo-framework/debug/dev/jovoSdk.js
/**
 * Created by Alex on 26-May-17.
 */
'use strict';

let request = require('request');
const API_URL = 'https://api.jovo.tech';

const JovoSdk = class {

    constructor(options) {
        this.token = options.token;
        this.skillId = options.skillId;
    }


    notification(title, body, userSkillId, callback) {
        let data = {
            skill_id: this.skillId,
            user_skill_id: userSkillId,
            command: 'notification',
            data: {
                notification: {
                    type: 'simple',
                    title: title,
                    body: body,
                },
            },
        };

        this.post('/android/command',
            data,
            callback);
    }

    data(data, userSkillId, callback) {
        let dataObj = {
            skill_id: this.skillId,
            user_skill_id: userSkillId,
            command: 'data',
            data: data,
        };

        this.post('/android/command',
            dataObj,
            callback);
    }

    post(path, data, callback) {
        request(
            {
                url: API_URL+path,
                headers: {
                    'content-type': 'application/json',
                    'Authorization': 'Bearer '+this.token,
                },
                method: 'POST',
                json: data,
            }, function(error, response, body) {
                callback(body);
            }
        );
    }


};


module.exports.JovoSdk = JovoSdk;
